### PR TITLE
Restrict pwm to channel

### PR DESCRIPTION
**Pull request by elionne**
Restrict PWM Channel to their corresponding pin.

Now you cannot specify a pin to the wrong PWN Channel.

### CODE REVIEW

**Review by Disasm — CHANGES_REQUESTED** [review]
Thank you!
I noticed that remap information is missing, both regular and remapped pins are used in the same way. This will not work unless you remap them explicitly outside of the driver. It's better to set this remap configuration in the driver itself to prevent related mistakes.

**Comment by elionne** [comment]
Hello,

Ok let me rework with remap feature.

File: src/pwm.rs [hunk 1]
```diff
@@ -6,85 +6,209 @@ use gd32vf103_pac::{TIMER0, TIMER1, TIMER2, TIMER3, TIMER4};
...
+        NoRemap: [ ch0: PA8, ch1: PA9, ch2: PA10, ch3: PA11 ],
+        PartialRemap1: [ ch0: PA8, ch1: PA9, ch2: PA10, ch3: PA11 ],
+        FullRemap: [ ch0: PE9, ch1: PE11, ch2: PE13, ch3: PE14 ],
+}
+
+pwm_pin!{
+    TIMER1,
+        NoRemap: [ch0: PA0, ch1: PA1, ch2: PA2, ch3: PA3 ],
+        PartialRemap1: [ch0: PA15, ch1: PB3, ch2: PA2, ch3: PA3 ],
+        PartialRemap2: [ch0: PA0, ch1: PA1, ch2: PB10, ch3: PB1 ],
```
Comment (Disasm) [inline src/pwm.rs:111]:
```suggestion
        PartialRemap2: [ch0: PA0, ch1: PA1, ch2: PB10, ch3: PB11 ],
```

File: src/pwm.rs [hunk 2]
```diff
@@ -126,28 +242,28 @@ macro_rules! general_pwm_timer {
...
-                    1 if self.ch1.is_some() => self.timer.chctl2.modify(|_r, w| w.ch1en().clear_bit()),
-                    2 if self.ch2.is_some() => self.timer.chctl2.modify(|_r, w| w.ch2en().clear_bit()),
-                    3 if self.ch3.is_some() => self.timer.chctl2.modify(|_r, w| w.ch3en().clear_bit()),
-                    _ => (),
+                match channel {
+                    Channel::CH0 => self.timer.chctl2.modify(|_r, w| w.ch0en().clear_bit()),
+                    Channel::CH1 => self.timer.chctl2.modify(|_r, w| w.ch1en().clear_bit()),
+                    Channel::CH2 => self.timer.chctl2.modify(|_r, w| w.ch2en().clear_bit()),
+                    Channel::CH3 => self.timer.chctl2.modify(|_r, w| w.ch3en().clear_bit()),
+                    Channel::NotUsed => {},
```
Comment (Disasm) [inline src/pwm.rs:256]: Hmm, this `NotUsed` variant doesn't look right here

**Review by Disasm — CHANGES_REQUESTED** [review]
Thank you! However, a few more changes are needed :)

I was also thinking if it's possible to make an API like this:
```rust
let pwm = PwmTimer::new(timer0, (None, Some(pa9), Some(pa10), None), &mut rcu, &mut afio);
```
This should be possible with a sort of `Pins<TIMER>` trait implemented for the tuple. Anyway, I'm happy to merge this PR as is after the fixes :)

**Comment by Disasm** [comment]
Oh, I've just found another problem:
```rust
let pwm_t0 = PwmConfiguration::<TIMER0, NoRemap>::new(dp.TIMER0, &mut afio);
let (_, ch1, ch2, _) = pwm_t0.pins(None, Some(&pa9), Some(&pa10), None);
let mut pwm_t0 = pwm_t0.configure(&mut rcu);

pa9.into_push_pull_output().set_high();
pwm_t0.set_duty(ch1, 100); // Oops!
```

**Comment by elionne** [comment]
Hello I finally did the rework with the suggest API. Now the code is simpler and API nicer. Thanks for that suggestion. Sorry for the delay.

**Comment by dkhayes117** [comment]
Interesting work, @elionne.  Looking forward for this to be merged.

**Review by dkhayes117 — APPROVED** [review]

**Comment by dkhayes117** [comment]
Thank you for your work!

**Comment by dkhayes117** [comment]
Oops, my mistake pushed the wrong button.

**Pull request merged**